src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java

```diff
@@ -92,14 +93,19 @@
      */
     public static final String MSG_KEY = "one.top.level.class";
 
+    /** Compare DetailAST nodes by line and then column number to make a unique ordering. */
+    private static final Comparator<DetailAST> LINE_AND_COL_COMPARATOR =
+            Comparator.comparingInt(DetailAST::getLineNo).thenComparingInt(DetailAST::getColumnNo);
```

Comment:
I am lost where the unique ordering is needed. Checkstyle will drop violations for the same line/column/message/check. It also already sorts violations on line/column. Can you explain what I am missing?

Reply:
The sorting is not so much for the order of the violations, but for this rule:
> If file doesn't contains public class, enum or interface, top-level type is the first type in file.

DetailAST on its own is not comparable, so we have to apply our own comparator to determine which is the first type in the file. This comes from extending the existing TreeMap to remember not just the line number, but the whole DetailAST node as well.

In hindsight, traversing the AST in `beginTree` should already visit the nodes in order. Maybe we can use a more efficient data structure instead? I'm thinking of using some sort of List; I just need a `List<DetailAST>` that stores the nodes in the traversed (sorted) order and I can get the individual type name from the node itself later.

Reply:
Thanks. I see now. Original map sorted on integer which is the key.

> Maybe we can use a more efficient data structure instead

Feel free to create a new issue and PR for this change.